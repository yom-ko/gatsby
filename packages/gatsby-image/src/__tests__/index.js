import "@babel/polyfill"
import React from "react"
import { render, cleanup, fireEvent } from "react-testing-library"
import Img from "../"

afterAll(cleanup)

const fixedShapeMock = {
  width: 100,
  height: 100,
  src: `test_image.jpg`,
  srcSet: `some srcSet`,
  srcSetWebp: `some srcSetWebp`,
  base64: `string_of_base64`,
}

const fluidShapeMock = {
  aspectRatio: 1.5,
  src: `test_image.jpg`,
  srcSet: `some srcSet`,
  srcSetWebp: `some srcSetWebp`,
  sizes: `(max-width: 600px) 100vw, 600px`,
  base64: `string_of_base64`,
}

const setup = (fluid = false, onLoad = () => {}, onError = () => {}) => {
  const { container } = render(
    <Img
      backgroundColor
      className={`fixedImage`}
      style={{ display: `inline` }}
      title={`Title for the image`}
      alt={`Alt text for the image`}
      {...fluid && { fluid: fluidShapeMock }}
      {...!fluid && { fixed: fixedShapeMock }}
      onLoad={onLoad}
      onError={onError}
      placeholderStyle={{ color: `red` }}
      placeholderClassName={`placeholder`}
    />
  )

  return container
}

describe(`<Img />`, () => {
  it(`should render fixed size images`, () => {
    const component = setup()
    expect(component).toMatchSnapshot()
  })

  it(`should render fluid images`, () => {
    const component = setup(true)
    expect(component).toMatchSnapshot()
  })

  it(`should have correct src, title and alt attributes`, () => {
    const imageTag = setup().querySelector(`picture img`)
    expect(imageTag.getAttribute(`src`)).toEqual(`test_image.jpg`)
    expect(imageTag.getAttribute(`title`)).toEqual(`Title for the image`)
    expect(imageTag.getAttribute(`alt`)).toEqual(`Alt text for the image`)
  })

  it(`should have correct placeholder src, title, style and class attributes`, () => {
    const placeholderImageTag = setup().querySelector(`img`)
    expect(placeholderImageTag.getAttribute(`src`)).toEqual(`string_of_base64`)
    expect(placeholderImageTag.getAttribute(`title`)).toEqual(
      `Title for the image`
    )
    // No Intersection Observer in JSDOM, so placeholder img will be visible (opacity 1) by default
    expect(placeholderImageTag.getAttribute(`style`)).toEqual(
      `position: absolute; top: 0px; left: 0px; width: 100%; height: 100%; opacity: 1; color: red;`
    )
    expect(placeholderImageTag.getAttribute(`class`)).toEqual(`placeholder`)
  })

  it(`should call onLoad and onError image events`, () => {
    const onLoadMock = jest.fn()
    const onErrorMock = jest.fn()
    const imageTag = setup(true, onLoadMock, onErrorMock).querySelector(
      `picture img`
    )
    fireEvent.load(imageTag)
    fireEvent.error(imageTag)

    expect(onLoadMock).toHaveBeenCalledTimes(1)
    expect(onErrorMock).toHaveBeenCalledTimes(1)
  })
})
