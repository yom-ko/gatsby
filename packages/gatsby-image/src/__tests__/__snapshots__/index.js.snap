// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Img /> should render fixed size images 1`] = `
<div>
  <div
    class="fixedImage gatsby-image-wrapper"
    style="position: relative; overflow: hidden; display: inline; width: 100px; height: 100px;"
  >
    <img
      alt=""
      class="placeholder"
      src="string_of_base64"
      style="position: absolute; top: 0px; left: 0px; width: 100%; height: 100%; opacity: 1; color: red;"
      title="Title for the image"
    />
    <div
      style="width: 100px; opacity: 1; height: 100px;"
      title="Title for the image"
    />
    <picture>
      <source
        srcset="some srcSetWebp"
        type="image/webp"
      />
      <source
        srcset="some srcSet"
      />
      <img
        alt="Alt text for the image"
        height="100"
        src="test_image.jpg"
        style="position: absolute; top: 0px; left: 0px; width: 100%; height: 100%; opacity: 0;"
        title="Title for the image"
        width="100"
      />
    </picture>
    <noscript>
      &lt;picture&gt;&lt;source type='image/webp' srcSet="some srcSetWebp" /&gt;&lt;source srcSet="some srcSet" /&gt;&lt;img width="100" height="100" src="test_image.jpg" alt="Alt text for the image" title="Title for the image" style="position:absolute;top:0;left:0;transition:opacity 0.5s;transition-delay:0.5s;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center"/&gt;&lt;/picture&gt;
    </noscript>
  </div>
</div>
`;

exports[`<Img /> should render fluid images 1`] = `
<div>
  <div
    class="fixedImage gatsby-image-wrapper"
    style="position: relative; overflow: hidden; display: inline;"
  >
    <div
      style="width: 100%; padding-bottom: 66.66666666666667%;"
    />
    <img
      alt=""
      class="placeholder"
      src="string_of_base64"
      style="position: absolute; top: 0px; left: 0px; width: 100%; height: 100%; opacity: 1; color: red;"
      title="Title for the image"
    />
    <div
      style="position: absolute; top: 0px; bottom: 0px; opacity: 1; right: 0px; left: 0px;"
      title="Title for the image"
    />
    <picture>
      <source
        sizes="(max-width: 600px) 100vw, 600px"
        srcset="some srcSetWebp"
        type="image/webp"
      />
      <source
        sizes="(max-width: 600px) 100vw, 600px"
        srcset="some srcSet"
      />
      <img
        alt="Alt text for the image"
        src="test_image.jpg"
        style="position: absolute; top: 0px; left: 0px; width: 100%; height: 100%; opacity: 0;"
        title="Title for the image"
      />
    </picture>
    <noscript>
      &lt;picture&gt;&lt;source type='image/webp' srcSet="some srcSetWebp" sizes="(max-width: 600px) 100vw, 600px" /&gt;&lt;source srcSet="some srcSet" sizes="(max-width: 600px) 100vw, 600px" /&gt;&lt;img src="test_image.jpg" alt="Alt text for the image" title="Title for the image" style="position:absolute;top:0;left:0;transition:opacity 0.5s;transition-delay:0.5s;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center"/&gt;&lt;/picture&gt;
    </noscript>
  </div>
</div>
`;
